### PR TITLE
Botcheck

### PR DESCRIPTION
**Pull request by magnumripper**

### CODE REVIEW

**Comment by solardiz** [comment]
On Thu, Sep 19, 2019 at 03:33:17PM -0700, magnum wrote:
>   * Make --users/groups/shells options case insensitive and allow wildcard

I think the "case insensitive" part is a misfeature.

**Comment by magnumripper** [comment]
> On Thu, Sep 19, 2019 at 03:33:17PM -0700, magnum wrote: * Make --users/groups/shells options case insensitive and allow wildcard
> I think the "case insensitive" part is a misfeature.

Depending on use case, I agree. So in order to keep this simple, how about we make it case insensitive only when a `*` wildcard is present?

**Comment by solardiz** [comment]
> how about we make it case insensitive only when a `*` wildcard is present?

On one hand, this is good enough when you know about that property. On the other hand, it's a subtle extra thing to learn, and most people would be confused by it at first. So I think we should either not implement the case-insensitivity at all or have it controlled separately from any other functionality. Even a john.conf setting would be better.

File: src/cracker.c [hunk 1]
```diff
@@ -477,11 +481,18 @@ static char *crk_loaded_counts(void)
...
+	                crk_db->password_count,
+	                crk_db->salt_count > 1 ?
+	                jtr_itoa(crk_db->salt_count, nbuf, 24, 10) : "no");
+
+	int b = (10 * crk_db->password_count + (crk_db->salt_count / 2)) /
+		crk_db->salt_count;
+
+	if (crk_db->salt_count > 1 && b > 10)
+		sprintf(s_loaded_counts + p, " (%d.%dx same-salt boost)",
+		        b / 10, b % 10);
```
Comment (solardiz) [inline src/cracker.c:495]: `sprintf` isn't supposed to fail, but formally it's defined to return a negative value on error. If the first `sprintf` call here fails, then the second one will use an out-of-bounds pointer. I think we should avoid this coding pattern, perhaps by checking that `p > 0` before attempting the second `sprintf`

File: src/john.c [hunk 2]
```diff
@@ -890,17 +890,24 @@ static void john_mpi_wait(void)
...
+	                database.password_count,
+	                database.salt_count > 1 ?
+	                jtr_itoa(database.salt_count, nbuf, 24, 10) : "no");
+
+	int b = (10 * database.password_count + (database.salt_count / 2)) /
+		database.salt_count;
+
+	if (database.salt_count > 1 && b > 10)
+		sprintf(s_loaded_counts + p, " (%d.%dx same-salt boost)",
+		        b / 10, b % 10);
```
Comment (solardiz) [inline src/john.c:910]: Ditto re: `sprintf` behavior on error. Also, let's avoid the code duplication.

File: src/cracker.c (hunk 1 above)
Comment (magnumripper) [reply src/cracker.c:495]: Good point, I'll look into that

**Comment by magnumripper** [comment]
> > how about we make it case insensitive only when a `*` wildcard is present?
> 
> On one hand, this is good enough when you know about that property. On the other hand, it's a subtle extra thing to learn, and most people would be confused by it at first. So I think we should either not implement the case-insensitivity at all or have it controlled separately from any other functionality. Even a john.conf setting would be better.

Until now, nothing in the docs mentioned it **is** case sensitive in the first place. The fact it was might perhaps have been expected for UNIX accounts but certainly not for eg. Windows accounts and many other ones. I, for one, always thought it was case insenstive.

Also, I want to keep this simple.

**Pull request merged**